src/Std/Data/Internal/List/Associative.lean

```diff
@@ -2999,6 +3005,469 @@ theorem isEmpty_insertList [BEq α]
+  exact getValueCast?_insertList_of_contains_left_eq_false distinct_l distinct_toInsert not_contains
+
+theorem length_insertList_distinct [BEq α] [EquivBEq α]
+    {l toInsert : List ((a : α) × β a)}
+    (distinct_l : DistinctKeys l)
+    (distinct_toInsert : DistinctKeys toInsert) :
+    (∀ (a : α), containsKey a l → containsKey a toInsert = false) →
+    (insertList l toInsert).length = l.length + toInsert.length := by
+    replace distinct_toInsert := DistinctKeys_impl_Pairwise_distinct distinct_toInsert
+    intro distinct_keys
+    have subgoal2 : ∀ (a : α), containsKey a l = true → (List.map Sigma.fst toInsert).contains a = false := by
+      intro a h_contains
+      specialize distinct_keys a h_contains
+      rwa [containsKey_eq_contains_map_fst] at distinct_keys
+    apply length_insertList distinct_l distinct_toInsert subgoal2
+
+theorem getKey?_insertList_of_mem_of_not_mem [BEq α] [LawfulBEq α]
+    {l toInsert : List ((a : α) × β a)} {k : α}
+    (contains : containsKey k l)
+    (not_contains : containsKey k toInsert = false) :
+    List.getKey? k (insertList l toInsert) = some k := by
+  simp only [getKey?_insertList_of_contains_left_of_contains_right_eq_false not_contains, getKey?_eq_some contains]
+
+theorem _root_.Option.or_eq_left_of_isSome {o o' : Option α} : o.isSome = true → o.or o' = o := by
+  cases o <;> simp
+
+theorem insertListIfNew_perm_insertList [BEq α] [EquivBEq α] {l₁ l₂ : List ((a : α) × β a)}
+    (hd₁ : DistinctKeys l₁) (hd₂ : DistinctKeys l₂) :
+    List.Perm (insertListIfNew l₁ l₂) (insertList l₂ l₁) := by
+  induction l₂ using assoc_induction generalizing l₁ with
+  | nil =>
+    simp only [insertListIfNew]
+    exact Perm.trans (by simp) (perm_insertList hd₂ (pairwise_fst_eq_false hd₁) (by simp)).symm
+  | cons k v l ih =>
+    simp only [insertListIfNew]
+    refine Perm.trans (ih hd₁.insertEntryIfNew hd₂.tail) ?_
+    refine getEntry?_ext hd₂.tail.insertList hd₂.insertList (fun k' => ?_)
+    rw [getEntry?_insertList hd₂.tail (pairwise_fst_eq_false hd₁.insertEntryIfNew),
+      getEntry?_insertList hd₂ (pairwise_fst_eq_false hd₁), getEntry?_insertEntryIfNew]
+    simp only [Bool.and_eq_true, Bool.not_eq_eq_eq_not, Bool.not_true]
+    split
+    · rename_i h
+      rw [Option.some_or, Option.or_eq_right_of_none, getEntry?_cons_of_true h.1]
+      rw [← getEntry?_congr h.1, getEntry?_eq_none, h.2]
+    · rename_i h
+      rw [getEntry?_cons]
+      by_cases hk : k == k'
+      · simp only [hk, true_and, Bool.not_eq_false] at h
+        suffices (getEntry? k' l₁).isSome by simp [Option.or_eq_left_of_isSome this]
+        rwa [← containsKey_eq_isSome_getEntry?, ← containsKey_congr hk]
+      · simp [hk]
+
+theorem insertList_perm_insertSmallerList [BEq α] [EquivBEq α]
+  {l toInsert : List ((a : α) × β a)}
+  (distinct_l : DistinctKeys l)
+  (distinct_toInsert : DistinctKeys toInsert) :
+  (insertList l toInsert).Perm
+  (insertSmallerList l toInsert) := by
+    unfold insertSmallerList
```

Comment:
Style: here you have the indentation the wrong way around: the statement should be indented four spaces, while the body should be indented two spaces.